Framework/Algorithms/src/UnwrapMonitor.cpp

```diff
@@ -109,23 +109,38 @@ void UnwrapMonitor::exec() {
       // If the detector flightpath is missing, zero the data
       g_log.debug() << "Detector information for workspace index " << i
                     << " is not available.\n";
-      tempWS->dataX(i).assign(tempWS->dataX(i).size(), 0.0);
-      tempWS->dataY(i).assign(tempWS->dataY(i).size(), 0.0);
-      tempWS->dataE(i).assign(tempWS->dataE(i).size(), 0.0);
+      tempWS->mutableX(i).assign(tempWS->x(i).size(), 0.0);
+      tempWS->mutableY(i).assign(tempWS->y(i).size(), 0.0);
+      tempWS->mutableE(i).assign(tempWS->e(i).size(), 0.0);
       continue;
     }
 
     // Unwrap the x data. Returns the bin ranges that end up being used
     const double Ld = L1 + spectrumInfo.l2(i);
-    const std::vector<int> rangeBounds = this->unwrapX(tempWS, i, Ld);
+    MantidVec newX;
+    const std::vector<int> rangeBounds = this->unwrapX(newX, i, Ld);
     // Unwrap the y & e data according to the ranges found above
-    this->unwrapYandE(tempWS, i, rangeBounds);
-    assert(tempWS->dataX(i).size() == tempWS->dataY(i).size() + 1);
+    MantidVec newY;
+    MantidVec newE;
+    this->unwrapYandE(tempWS, i, rangeBounds, newY, newE);
```

Comment:
I am not happy either, but the way the algorithm is implemented conflicts with the immutable Histogram concept. 

I had thought about combining the methods earlier, but I rejected it for two reasons: 
1. We don't have testing on this algorithm, hence larger changes are risky. 
2. Combining the two methods essentially means having the code of two methods in one.

We could have a governing method which wraps these two steps. In combination with documentation this might be sufficient.